api/src/main/scala/de/zalando/play/controllers/SecurityExtractors.scala

```diff
@@ -0,0 +1,129 @@
+package de.zalando.play.controllers
+
+import java.net.URLEncoder
+
+import play.api.libs.json.JsValue
+import play.api.mvc.{Result, ActionBuilder, Request, RequestHeader}
+import play.api.mvc.Security.AuthenticatedRequest
+import sun.misc.BASE64Decoder
+
+import scala.concurrent.Future
+import play.api.libs.iteratee.Execution.Implicits.trampoline
+
+import scala.language.implicitConversions
+
+
+/**
+  * @author slasch 
+  * @since 07.03.2016.
+  *
+  * Helper class for long authentication process
+  * @param userinfo
+  * @param onUnauthorized
+  * @tparam U
+  */
+class FutureAuthenticatedBuilder[U](userinfo: RequestHeader => Future[Option[U]],
+                              onUnauthorized: RequestHeader => Result)
+  extends ActionBuilder[({ type R[A] = AuthenticatedRequest[A, U] })#R] {
+
+  override def invokeBlock[A](request: Request[A], block: (AuthenticatedRequest[A, U]) => Future[Result]) =
+    authenticate(request, block)
+
+  def authenticate[A](request: Request[A], block: (AuthenticatedRequest[A, U]) => Future[Result]) = {
+    userinfo(request).flatMap { userOpt =>
+      userOpt.map { user =>
+        block(new AuthenticatedRequest(user, request))
+      } getOrElse {
+        Future.successful(onUnauthorized(request))
+      }
+    }
+  }
+}
+
+
+object SwaggerSecurityExtractors extends BasicAuthSecurityExtractor with OAuthResourceOwnerPasswordCredentialsFlow {
+  private implicit def option2futureOption[User](o:Option[User]): Future[Option[User]] = Future.successful(o)
+
+  def basicAuth[User >: Any]: RequestHeader => ((String, String) => User) => Future[Option[User]] =
+    header => convertToUser => header.headers.get("Authorization").map { basicAuth =>
+      decodeBasicAuth(basicAuth).map(p => convertToUser(p._1, p._2))
```

Comment:
Could you use maybe use 
`map{case(username, password) => convertToUser(username, password)}`
here? Would look nicer